homeassistant/components/device_tracker/owntracks.py

```diff
@@ -60,47 +67,101 @@ def owntracks_event_update(topic, payload, qos):
             data = json.loads(payload)
         except ValueError:
             # If invalid JSON
-            logging.getLogger(__name__).error(
+            _LOGGER.error(
                 'Unable to parse payload as JSON: %s', payload)
             return
 
         if not isinstance(data, dict) or data.get('_type') != 'transition':
             return
 
-        # check if in "home" fence or other zone
-        location = ''
+        if data['desc'].lower() == 'home':
+            location = STATE_HOME
+        else:
+            location = data['desc']
+
+        dev_id, kwargs = _parse_see_args(topic, data)
+
         if data['event'] == 'enter':
+            zone = hass.states.get("zone.{}".format(location))
+            with LOCK:
+                if zone is not None:
+                    kwargs['location_name'] = location
 
-            if data['desc'].lower() == 'home':
-                location = STATE_HOME
-            else:
-                location = data['desc']
+                    regions = REGIONS_ENTERED[dev_id]
+                    if location not in regions:
+                        regions.append(location)
+                    _LOGGER.info("Enter region %s", location)
+                    _get_gps_from_zone(kwargs, zone)
+
+                see(**kwargs)
 
         elif data['event'] == 'leave':
-            location = STATE_NOT_HOME
+            current_location = hass.states.get(
+                "device_tracker.{}".format(dev_id)).state
+
+            regions = REGIONS_ENTERED[dev_id]
+            if location in regions:
+                regions.remove(location)
+            new_region = regions[-1] if regions else None
+
+            if not new_region:
+                _LOGGER.info("Exit from %s to GPS", location)
+                see(**kwargs)
+
+                # if gps location didn't set new state, force to away
+                current_location = hass.states.get(
```

Comment:
Less of a hack than what was there before! (Which was to always set STATE_NOT_HOME whenever you exited a region! )

What I think the previous author was trying to do was to cope with the situation where you get a zone exit on the boundary of a zone. You should respect that OwnTracks has said the zone was exited.  However if the GPS location is on the boundary HA may stick you back in the zone.

So what I've tried to do is allow HA to respond sensibly - and fallback to what the old code did, if HA didn't exit the zone.

My experiments indicate this is questionable using beacons - because they can randomly disconnect -  if that happens a forced exit is probably wrong. So could just do this for gps exits.

If you want to keep comparability with the old code we need some sort of forced exit.

Purity is perhaps a suggestion to make your HA zone smaller than your OwnTracks one, and remove the forced exit.

I don't think there is a perfect solution here -we're trying to reconcile two incompatible data sources. So happy to go with whatever makes most sense.


As you probably know the issue is that the OwnTracks location tracking can have long gaps between updates. So if you incorrectly interpret an exit - the state could be wrong for a long time.